### ISSUE TITLE
Reduce dependencies

### ISSUE THREAD

**Issue by Brian-McBride** [swantzter/apollo-datasource-firestore#9]
To make the datasource even more useful outside of Apollo, you could probably drop the actual apollo imports and just create a few types

### Datasource
```typescript
export interface DataSourceConfig<TContext> {
  context: TContext;
  cache: KeyValueCache;
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export abstract class DataSource<TContext = any> {
  initialize?(config: DataSourceConfig<TContext>): void | Promise<void>;
}
```

### Cache
```typescript
export interface KeyValueCacheSetOptions {
  /**
   * Specified in **seconds**, the time-to-live (TTL) value limits the lifespan
   * of the data being stored in the cache.
   */
  ttl?: number | null;
}

export interface KeyValueCache<V = string> {
  get(key: string): Promise<V | undefined>;
  set(key: string, value: V, options?: KeyValueCacheSetOptions): Promise<void>;
  delete(key: string): Promise<boolean | void>;
}

export class InMemoryLRUCache<V = string> implements KeyValueCache<V> {
  private store: LRUCache<string, V>;

  // TODO: Define reasonable default max size of the cache
  constructor({
    maxSize = Infinity,
    sizeCalculator = defaultLengthCalculation,
    onDispose,
  }: {
    maxSize?: number;
    sizeCalculator?: (value: V, key: string) => number;
    onDispose?: (key: string, value: V) => void;
  } = {}) {
    this.store = new LRUCache({
      max: maxSize,
      length: sizeCalculator,
      dispose: onDispose,
    });
  }

  async get(key: string) {
    return this.store.get(key);
  }
  async set(key: string, value: V, options?: { ttl?: number }) {
    const maxAge = options?.ttl && options.ttl * 1000;
    this.store.set(key, value, maxAge);
  }
  async delete(key: string) {
    this.store.del(key);
  }

  // Drops all data from the cache. This should only be used by test suites ---
  // production code should never drop all data from an end user cache.
  async flush(): Promise<void> {
    this.store.reset();
  }
  async getTotalSize() {
    return this.store.length;
  }

  // This is a size calculator based on the number of bytes in a JSON
  // encoding of the stored object. It happens to be what ApolloServer
  // uses for its default DocumentStore and may be helpful to others as well.
  static jsonBytesSizeCalculator<T>(obj: T): number {
    return Buffer.byteLength(JSON.stringify(obj), 'utf8');
  }
}

function defaultLengthCalculation(item: any) {
  if (Array.isArray(item) || typeof item === 'string') {
    return item.length;
  }

  // Go with the lru-cache default "naive" size, in lieu anything better:
  //   https://github.com/isaacs/node-lru-cache/blob/a71be6cd/index.js#L17
  return 1;
}
```

Then just change the error line to 
```typescript
      throw new Error(
        'FirestoreDataSource must be created with a Firestore collection (from @google-cloud/firestore)'
      );
```

Of course, I used the code from Apollo for the cache. It does make me wonder that if no cache is supplied, should one be created inside the class? Or maybe it should just be a non-functional cache?
```typescript
const noCache = {
  get: (key: string) => Promise.resolve(undefined),
  set: (key: string) => Promise.resolve(undefined),
  delete: (key: string) => Promise.resolve(undefined),
};
```

There is a good argument for making the cache very declarative.

**Comment by swantzter** [comment]
I can see the appeal, but it would lose a lot in terms of maintainability for me. The point of depending on those packages for types is that I can easily make sure the package is compatible with apollo-server, its intended usecase, by keeping those packages up to date and being yelled at by typescript.

The package is perfectly useful without apollo as-is, from my experience, the largest issue I suppose it that it has a sub-dependency on graphql (2.7M) that bumps up the size, that comes from `apollo-server-errors` which, to your point, might be worth removing

**Comment by Brian-McBride** [comment]
You can still import the Apollo packages and use them in testing. 
That would give you any fail results in the rare instance that the interface for the Apollo Datasource actually changed. 

If that happened, it would be a major breaking change anyway and probably require changes to the datasource in general. 

I only mention it as I'm finding a use for dataloaders in some non-graphql cases as well. Having a nice batch/cache datasource is handy even in some more standard RESTful apis. The size savings is nice too. I didn't think of that as much as just being able to use it outside of Apollo.

**Comment by swantzter** [comment]
I've released [v5.2.0](https://github.com/swantzter/apollo-datasource-firestore/releases/tag/v5.2.0) which removes the dependency on `apollo-server-errors`

Yes, if the underlying datasource types change in a major way this library would need a major version, that's already happened once.

I'm getting a lot of value as a maintainer by depending on the remaining apollo packages which mainly provides typings, I get a clear indication when changes are needed, I don't have to maintain a copy of the "API" that the library has a goal to adhere to, and I can "re-export" the types for better typings for dependants.
It's orders of magnitude simpler from a maintainer point of view, and at this point in time I don't have the availability to make things less practical  to maintain, especially as I don't see what would be gained.

I don't really understand your point that inlining these type definitions/using another default cache implementation would make it easier to use the datasource in non-graphql applications - I use this in several non-graphql/non-apollo projects without any issues. Am I missing some specific usecase where it creates problems?

**Comment by Brian-McBride** [comment]
## Datasource Types

No, it won't cause problems. 

But if you are building a RESTful interface or maybe a gRPC, you'll end up pulling in the npm modules for Apollo and then GraphQL and so on. It is only based on reducing npm imports.

But again, you can still use the Apollo libs in testing. If you do typescript linting/compiling it will throw a type error if the values don't align. If the interface for Datasource was more complicated, it wouldn't be worth it for sure. As it is so crazy simple...

Also, it doesn't need to be called Datasource. It could be DatasourceRef or whatever. And, it also wouldn't really need to be exported, but it is handy if you are working in an environment that is completely devoid of Apollo. 

Yet again - this is just being efficient with NPM imports. Honestly, it makes no difference operationally. 

-----

## Cache 

The actual cache interface is also a copy from Apollo. Just not pulling in the Apollo npm modules in a project they are not needed. This is pretty much a copy/paste from the Apollo source.
```typescript
export interface KeyValueCacheSetOptions {
  /**
   * Specified in **seconds**, the time-to-live (TTL) value limits the lifespan
   * of the data being stored in the cache.
   */
  ttl?: number | null;
}

export interface KeyValueCache<V = string> {
  get(key: string): Promise<V | undefined>;
  set(key: string, value: V, options?: KeyValueCacheSetOptions): Promise<void>;
  delete(key: string): Promise<boolean | void>;
}
```

The other stuff I put in above was the actual code Apollo uses for InMemoryLRUCache. Which, is really a default wrapper around LRUcache - but converted to async calls. Mainly this allows easier interop between async cache sources like Redis. 

When I suggested using a noCache
```typescript
const noCache = {
  get: (key: string) => Promise.resolve(undefined),
  set: (key: string) => Promise.resolve(undefined),
  delete: (key: string) => Promise.resolve(undefined),
};
```
This would then change the line in initialize to:
`cache: cache ?? noCache,`

This is just a choice between being declarative or not. Should an LRUcache be auto-created if a user didn't specify one? But then, what are the settings (like max size)? Could that cause debug confusion? Should a developer have to actually do the work to specify and create their cache with settings that match their project to have caching enabled at all? Or should every datasource just automatically start something that is going to take up an unknown memory footprint? 

It is similar to ending your lines in semicolons or not. In the past, I've built my datasources exactly like you have where if Apollo doesn't give a cache then I just create one myself. I ran across a project where this was done all over the place and they were killing their NodeJS ram limit with a mixture of heavy load and large data payloads. Once we specified a central cache (or at least defined them declaratively) the issue went away. 

-----

To be clear though, these are just freely given ideas and they are worth exactly that. I super appreciate your contribution to the community. 

I have some ideas about sub-collections as well. I expanded on the Mongo datasource to use the dataloader on complex queries instead of only ID fetches. I ended up using a hash of the query for a cache reference. I bet something similar could be done with sub-collections. Firestore is a bit more challenging as Mongo can really combine a lot in a big `$or [...]` wrapper.
